File: bai2.py
#!/usr/bin/env python3
"""
Bài 2: Xác minh chữ ký bằng khóa công khai
Mô tả: Ký một thông điệp bằng khóa bí mật RSA
Yêu cầu:
• Đọc nội dung từ file văn bản message.txt
• Ký nội dung bằng thuật toán SHA-256 với khóa bí mật
• Lưu chữ ký vào file signature.bin
"""

from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import serialization, hashes
import os
import sys

def read_message(filename):
    """Read message from text file"""
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            content = f.read()
        return content
    except FileNotFoundError:
        print(f"[ERROR] File not found: {filename}")
        return None
    except Exception as e:
        print(f"[ERROR] Error reading file: {e}")
        return None

def load_private_key(filename):
    """Load RSA private key from PEM file"""
    try:
        with open(filename, 'rb') as f:
            private_key = serialization.load_pem_private_key(
                f.read(),
                password=None
            )
        return private_key
    except FileNotFoundError:
        print(f"[ERROR] Private key file not found: {filename}")
        return None
    except Exception as e:
        print(f"[ERROR] Error loading private key: {e}")
        return None

def sign_message(message, private_key):
    """Sign message using RSA private key with SHA-256"""
    try:
        signature = private_key.sign(
            message.encode('utf-8'),
            padding.PSS(
                mgf=padding.MGF1(hashes.SHA256()),
                salt_length=padding.PSS.MAX_LENGTH
            ),
            hashes.SHA256()
        )
        return signature
    except Exception as e:
        print(f"[ERROR] Error signing message: {e}")
        return None

def save_signature(signature, filename):
    """Save signature to binary file"""
    try:
        with open(filename, 'wb') as f:
            f.write(signature)
        return True
    except Exception as e:
        print(f"[ERROR] Error saving signature: {e}")
        return False

def main():
    print("=" * 50)
    print("  RSA Digital Signature (SHA-256)")
    print("  Buoi 6 - Bai 2")
    print("=" * 50)
    print()

    # Configuration
    message_file = "message.txt"
    private_key_file = "private_key.pem"
    signature_file = "signature.bin"

    print("[*] Configuration:")
    print(f"    Message file: {message_file}")
    print(f"    Private key file: {private_key_file}")
    print(f"    Signature output: {signature_file}")
    print()

    # Read message
    print(f"[*] Reading message from {message_file}")
    message = read_message(message_file)
    if message is None:
        print("[ERROR] Failed to read message!")
        return False
    print(f"[+] Message read successfully ({len(message)} bytes)")
    print(f"[+] Message content: {message}")
    print()

    # Load private key
    print(f"[*] Loading private key from {private_key_file}")
    private_key = load_private_key(private_key_file)
    if private_key is None:
        print("[ERROR] Failed to load private key!")
        return False
    print("[+] Private key loaded successfully")
    print()

    # Sign message
    print("[*] Signing message with RSA private key (SHA-256)...")
    signature = sign_message(message, private_key)
    if signature is None:
        print("[ERROR] Failed to sign message!")
        return False
    print(f"[+] Message signed successfully ({len(signature)} bytes)")
    print()

    # Save signature
    print(f"[*] Saving signature to {signature_file}")
    if not save_signature(signature, signature_file):
        print("[ERROR] Failed to save signature!")
        return False
    print(f"[+] Signature saved successfully")
    print()

    # Print signature in hex format
    print("[+] Signature (hex):")
    print(f"    {signature.hex()}")
    print()

    # Print file information
    sig_size = os.path.getsize(signature_file)
    print("[+] File information:")
    print(f"    Message size: {len(message)} bytes")
    print(f"    Signature size: {sig_size} bytes")
    print()

    print("[SUCCESS] Message signed and signature saved!")
    return True

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: bai2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/sha.h>
#include <openssl/err.h>
#include <openssl/evp.h>
#include <cstring>

using namespace std;

// Function to read file content
string readFile(const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file.is_open()) {
        cerr << "Error: Cannot open file " << filename << endl;
        return "";
    }
    
    string content((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
    file.close();
    return content;
}

// Function to write binary file
bool writeFile(const string& filename, const unsigned char* data, int length) {
    ofstream file(filename, ios::binary);
    if (!file.is_open()) {
        cerr << "Error: Cannot create file " << filename << endl;
        return false;
    }
    
    file.write(reinterpret_cast<const char*>(data), length);
    file.close();
    return true;
}

// Function to load RSA private key from PEM file
RSA* loadPrivateKey(const string& keyFile) {
    FILE* fp = fopen(keyFile.c_str(), "r");
    if (!fp) {
        cerr << "Error: Cannot open key file " << keyFile << endl;
        return nullptr;
    }
    
    RSA* rsa = PEM_read_RSAPrivateKey(fp, nullptr, nullptr, nullptr);
    fclose(fp);
    
    if (!rsa) {
        cerr << "Error: Cannot read RSA private key" << endl;
        ERR_print_errors_fp(stderr);
        return nullptr;
    }
    
    return rsa;
}

// Function to compute SHA-256 hash
unsigned char* computeSHA256(const string& message, unsigned int& digest_len) {
    unsigned char* digest = new unsigned char[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, message.c_str(), message.length());
    SHA256_Final(digest, &sha256);
    
    digest_len = SHA256_DIGEST_LENGTH;
    return digest;
}

// Function to sign message using RSA
bool signMessage(const string& messageFile, const string& keyFile, const string& signatureFile) {
    // Read message
    cout << "[*] Reading message from " << messageFile << endl;
    string message = readFile(messageFile);
    if (message.empty()) {
        return false;
    }
    cout << "[+] Message read successfully (" << message.length() << " bytes)" << endl;
    cout << "[+] Message content: " << message << endl;
    
    // Load private key
    cout << "[*] Loading private key from " << keyFile << endl;
    RSA* rsa = loadPrivateKey(keyFile);
    if (!rsa) {
        return false;
    }
    cout << "[+] Private key loaded successfully" << endl;
    
    // Compute SHA-256 hash
    cout << "[*] Computing SHA-256 hash of message..." << endl;
    unsigned int digest_len;
    unsigned char* digest = computeSHA256(message, digest_len);
    cout << "[+] SHA-256 hash computed (" << digest_len << " bytes)" << endl;
    
    // Sign the hash
    cout << "[*] Signing hash with RSA private key..." << endl;
    unsigned char* signature = new unsigned char[RSA_size(rsa)];
    unsigned int signature_len;
    
    // Use RSA_sign with SHA256
    if (RSA_sign(NID_sha256, digest, digest_len, signature, &signature_len, rsa) != 1) {
        cerr << "Error: RSA_sign failed" << endl;
        ERR_print_errors_fp(stderr);
        delete[] digest;
        delete[] signature;
        RSA_free(rsa);
        return false;
    }
    cout << "[+] Message signed successfully (" << signature_len << " bytes)" << endl;
    
    // Save signature to file
    cout << "[*] Saving signature to " << signatureFile << endl;
    if (!writeFile(signatureFile, signature, signature_len)) {
        delete[] digest;
        delete[] signature;
        RSA_free(rsa);
        return false;
    }
    cout << "[+] Signature saved successfully" << endl;
    
    // Print signature in hex format
    cout << "\n[+] Signature (hex): ";
    for (unsigned int i = 0; i < signature_len; i++) {
        printf("%02x", signature[i]);
    }
    cout << "\n" << endl;
    
    // Cleanup
    delete[] digest;
    delete[] signature;
    RSA_free(rsa);
    
    return true;
}

int main() {
    // Initialize OpenSSL
    ERR_load_crypto_strings();
    OpenSSL_add_all_algorithms();
    
    cout << "========================================" << endl;
    cout << "  RSA Digital Signature (SHA-256)" << endl;
    cout << "  Buoi 6 - Bai 2" << endl;
    cout << "========================================\n" << endl;
    
    string messageFile = "message.txt";
    string keyFile = "private_key.pem";
    string signatureFile = "signature.bin";
    
    cout << "[*] Configuration:" << endl;
    cout << "    Message file: " << messageFile << endl;
    cout << "    Private key file: " << keyFile << endl;
    cout << "    Signature output: " << signatureFile << endl << endl;
    
    // Sign the message
    if (signMessage(messageFile, keyFile, signatureFile)) {
        cout << "[SUCCESS] Message signed and signature saved!" << endl;
        return 0;
    } else {
        cout << "[ERROR] Failed to sign message!" << endl;
        return 1;
    }
}
